components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/OrientAsyncHelper.java

```diff
@@ -0,0 +1,162 @@
+   * and returns a true asynchronous backed {@link Iterable} of {@link ODocument} instance, that can be later
+   * transformed into whatever caller needs. This means that this wrapper supports ANY SQL commands, even those
+   * performing aggregation on results, as long as iterator is not accumulated into a real array or list. Still, given
+   * the nature of non-blocking query, it's execution happens on a separate thread, hence, it will not see uncommited
+   * changes from the caller thread, if any. This method allows to tune buffer queue size and timeout.
+   */
+  public static Iterable<ODocument> asyncIterable(final ODatabaseDocumentTx db,
+                                                  final String selectQuery,
+                                                  @Nullable final Map<String, Object> parameters,
+                                                  final int bufferSize,
+                                                  final long timeoutSeconds)
+  {
+    checkNotNull(db);
+    checkNotNull(selectQuery);
+    checkArgument(bufferSize > 0);
+    checkArgument(timeoutSeconds >= 0);
+    final BlockingQueue<ODocument> queue = new ArrayBlockingQueue<>(bufferSize);
+
+    db.command(new OSQLNonBlockingQuery<ODocument>(selectQuery, new OCommandResultListener()
+    {
+      @Override
+      public boolean result(final Object iRecord) {
+        final ODocument doc = (ODocument) iRecord;
+        try {
+          queue.offer(doc, timeoutSeconds, TimeUnit.SECONDS);
+        }
+        catch (InterruptedException e) {
+          log.warn("Interrupted result", e);
+          return false;
+        }
+        return true;
+      }
+
+      @Override
+      public void end() {
+        try {
+          queue.offer(SENTINEL, timeoutSeconds, TimeUnit.SECONDS);
+        }
+        catch (InterruptedException e) {
+          log.warn("Interrupted end", e);
+        }
+      }
+    })).execute(parameters);
+
+    return new OrientAsyncIterableWrapper(timeoutSeconds, queue);
+  }
+
+  private static final ODocument SENTINEL = new ODocument();
+
+  private static final class OrientAsyncIterableWrapper
+      implements Iterable<ODocument>, Iterator<ODocument>
+  {
+    private final long timeoutSeconds;
+
+    private final BlockingQueue<ODocument> queue;
+
+    private ODocument next;
+
+    public OrientAsyncIterableWrapper(final long timeoutSeconds, final BlockingQueue<ODocument> queue)
+    {
+      this.timeoutSeconds = timeoutSeconds;
+      this.queue = queue;
+    }
+
+    @Override
+    public Iterator<ODocument> iterator() {
+      return this;
+    }
+
+    @Override
+    public boolean hasNext() {
+      if (next == null) {
+        try {
+          next = queue.poll(timeoutSeconds, TimeUnit.SECONDS);
+        }
+        catch (InterruptedException e) {
+          throw Throwables.propagate(e);
+        }
+      }
+      return next != SENTINEL;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Use equals() to compare object references. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=pmd%3ACompareObjectsWithEquals)